### PR TITLE
Add a JWT based auth context provider

### PR DESCRIPTION
**Pull request by Cito**
Proposal for a univeral JWT auth context provider.

This PR includes an example app that shows how this can be used with FastAPI,
an how the provider and the app can be set up by injecting a configuration (or any other dependencies).

We may need to reconsider whether this really should be part of hexkit or put elsewhere,
since it is more like a utility than a provider in the sense of hexkit, and no translator either
(the auth.connector module could be considered the translator).

This still needs some unit tests for the provider and e2e tests for the demo applicaton.

Update: It was decided to move this to the ghga-service-commons library.

### CODE REVIEW

File: examples/auth_demo/auth_demo/main.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+
+async def run_server():
+    """Run the HTTP API."""
+    config = Config()
+    async with get_configured_container(config) as container:
+        container.wire(modules=["auth_demo.router", "auth_demo.auth.connector"])
+        uv_config = uvicorn.Config(app=get_app())
+        server = uvicorn.Server(uv_config)
+        await server.serve()
```
Comment (KerstenBreuer) [inline examples/auth_demo/auth_demo/main.py:51]: Maybe better use the run function provided by the chassis lib: https://github.com/ghga-de/download-controller-service/blob/main/dcs/main.py#L67

**Review by KerstenBreuer — CHANGES_REQUESTED** [review]
Very nice. Just some small comments. Maybe it makes in deed sense to move this into a dedicated library as it does not touch the core of applications. Let's discuss.

File: examples/auth_demo/auth_demo/main.py (hunk 1 above)
Comment (Cito) [reply examples/auth_demo/auth_demo/main.py:51]: Didn't want to introduce another dependency for the example. Let's discuss where to put the whole thing first.

File: hexkit/protocols/auth.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    "AuthContextValidationError",
+]
+
+
+class AuthContextError(RuntimeError):
+    """Error when retrieving the authentication and authorization context failed."""
+
+
+class AuthContextValidationError(RuntimeError):
+    """Error that is raised when the underlying token is invalid."""
```
[the remark this answers is not in the document]
Comment (Cito) [reply hexkit/protocols/auth.py]: Changed in a1cba05b84a95635bf48f497f5cfed7495e1783f.

File: hexkit/providers/jwtauth/provider.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    )
+    auth_check_claims: dict[str, Any] = Field(
+        dict.fromkeys("name email iat exp".split()),
+        description="A dict of all claims that shall be verified by the provider."
+        + " A value of None means that the claim can have any value.",
+    )
+    auth_map_claims: dict[str, str] = Field(
+        {},
+        description="A mapping of claims to attributes in the auth context."
+        + " Only differently named attributes must be specified."
```
[the remark this answers is not in the document]
Comment (Cito) [reply hexkit/providers/jwtauth/provider.py]: Changed in 7409701255885ac38b828e0655cde0849e3eb5ee

File: hexkit/providers/jwtauth/provider.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        for claim, attribute in self._map_claims.items():
+            try:
+                value = claims.pop(claim)
+            except KeyError as exc:
+                raise JWTAuthValidationError(f"Missing claim {claim}") from exc
+            if attribute is not None:
+                claims[attribute] = value
+        try:
+            claims["expires"] = "Indemann"
+            return self._context_class(**claims)
```
[the remark this answers is not in the document]
Comment (Cito) [reply hexkit/providers/jwtauth/provider.py]: This is included in 32814c6958f9821bb7556b7a3166ff4b2e50a3fb.

File: examples/auth_demo/auth_demo/main.py (hunk 1 above)
Comment (KerstenBreuer) [reply examples/auth_demo/auth_demo/main.py:51]: Makes sense.

File: hexkit/providers/jwtauth/provider.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            UnicodeDecodeError,
+            KeyError,
+            TypeError,
+            ValueError,
+        ) as exc:
+            raise JWTAuthValidationError(f"Not a valid token: {exc}") from exc
+        try:
+            claims = json.loads(jwt_token.claims)
+        except json.JSONDecodeError as exc:
+            raise JWTAuthValidationError(f"Claims cannot be decoded: {exc}") from exc
```
[the remark this answers is not in the document]
Comment (Cito) [reply hexkit/providers/jwtauth/provider.py]: Had changed this already somewhere.

**Review by KerstenBreuer — COMMENTED** [review]
I think this looks good. However, should we close this PR, since we will move it over to the ghga_service_commons repo?

**Comment by Cito** [comment]
Yes, I will close this and delete the branch once it's moved.

**Comment by Cito** [comment]
Sorry, yes, should have closed this already.

**Pull request closed without merging**